File: py/candlestick_patterns/patterns/spinning_top.py
"""Spinning Top pattern (1-candle)."""
from __future__ import annotations

from ..core.primitives import is_white, is_black, real_body, upper_shadow, lower_shadow
from ...fuzzy import t_product_all


def spinning_top(self) -> float:
    """Spinning Top: a one-candle pattern.

    A candle with a small body and shadows longer than the body on both sides.

    The meaning of "short" is specified with self._short_body.

    Category C: color determines sign.

    Returns:
        Continuous float in [-100, +100].
    """
    if not self._enough(1, self._short_body):
        return 0.0

    o, h, l, c = self._bar(1)

    rb = real_body(o, c)

    mu_short = self._mu_less(rb, self._short_body, 1)

    # Shadows > body: positional comparisons.
    us = upper_shadow(o, h, c)
    ls = lower_shadow(o, l, c)
    width_us = self._fuzz_ratio * rb if rb > 0.0 else 0.0
    width_ls = self._fuzz_ratio * rb if rb > 0.0 else 0.0
    mu_us_gt_rb = self._mu_gt_raw(us, rb, width_us)
    mu_ls_gt_rb = self._mu_gt_raw(ls, rb, width_ls)

    confidence = t_product_all(mu_short, mu_us_gt_rb, mu_ls_gt_rb)

    if is_white(o, c):
        return confidence * 100.0
    if is_black(o, c):
        return -confidence * 100.0
    return 0.0

File: ts/candlestick-patterns/patterns/spinning-top.ts
/** Spinning Top pattern (1-candle). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll } from '../../fuzzy/index.ts';
import { realBody, upperShadow, lowerShadow, isWhite, isBlack } from '../core/primitives.ts';

/**
 * Spinning Top: a one-candle pattern.
 *
 * A candle with a small body and shadows longer than the body on both sides.
 *
 * The meaning of "short" is specified with `shortBody`.
 *
 * Category C: color determines sign.
 *
 * Returns:
 *     Continuous float in [-100, +100].
 */
export function spinningTop(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(1, cp.shortBody)) return 0.0;

    const b = cp.bar(1);
    const rb = realBody(b.o, b.c);

    const muShort = cp.muLessCS(rb, cp.shortBody, 1);

    // Shadows > body: positional comparisons.
    const us = upperShadow(b.o, b.h, b.c);
    const ls = lowerShadow(b.o, b.l, b.c);
    let widthUS = 0.0;
    if (rb > 0.0) widthUS = cp.fuzzRatio * rb;
    let widthLS = 0.0;
    if (rb > 0.0) widthLS = cp.fuzzRatio * rb;
    const muUSGtRB = cp.muGtRaw(us, rb, widthUS);
    const muLSGtRB = cp.muGtRaw(ls, rb, widthLS);

    const confidence = tProductAll(muShort, muUSGtRB, muLSGtRB);

    if (isWhite(b.o, b.c)) return confidence * 100.0;
    if (isBlack(b.o, b.c)) return -confidence * 100.0;
    return 0.0;
}
